README.md
# Custom Error Pages

Create custom `401` and `403` error pages with any **WordPress** theme without
writing a single line of code, set it up and forget it.

## Description

**WordPress** offers inbuilt support for custom `404` pages on all themes. But 
what about custom pages for other common errors like `401` and `403`? You end up
seeing a bland error page of the Web Server.

With this plugin you can easily create custom `401` and `403` error pages with
any theme without writing a single line of code!!! And the best part is that you
set it and forget it. Yes, you don't have to do any changes even if you change
themes. The heading and text you want on `401` and `403` error pages are
displayed on the currently active theme.

## Further Reading
* [Create WordPress 401 and 403 error pages](https://websistent.com/WordPress-custom-403-401-error-page/) WITHOUT using this plugin.
* The [Custom Error Pages Plugin](https://websistent.com/WordPress-plugins/custom-error-pages/) official homepage.

## Installation

1. Install and activate the custom error pages plugin.
2. From the **wp-admin** go to *Settings > Custom Error Pages*, fill in the
   heading and content boxes, save the changes and preview it.
3. Configure your web server to use a custom error page.

Apache users edit your `.htaccess` file and add the following lines.

	ErrorDocument 403 /index.php?status=403
	ErrorDocument 401 /index.php?status=401

Nginx users edit your `nginx.conf` file and add the following lines.

	error_page	403 = /index.php?status=403
	error_page	401 = /index.php?status=401

Try accessing a file or directory which is forbidden like

http://www.example.com/.htaccess

http://www.example.com/wp-includes/

Did you see a custom `403` page in all the glory of your theme?

## Frequently Asked Questions

### Should I do anything if I change themes?
No not at all. You just install the plugin set it and forget it. The plugin
isn't theme bound so it'll work even after switching themes.

### I noticed define( 'DONOTCACHEPAGE', TRUE ); in the plugin code what is it for?
This tells caching plugins like W3 Total Cache and WP Super Cache not to cache
the `401` and `403` error pages. This does **NOT** affect the cacheability any
other pages of your site.

### Does this plugin automatically edit the .htaccess or nginx.conf file?
No it doesn't. You have to manually edit it and add the necessary configuration
specified in the plugin's option page.

## Credits

This plugin was written by @jesinwp.

Version `1.2` in that repository was made by @sizious, so it's unofficial.


CHANGELOG.md
# Changelog

## 1.2

* Fixed a bug that displayed all comments of the WordPress database on the generated error page.
* Added the translation of the plugin in "French (France)" language.

## 1.1

* Fixed a bug that displayed the thumbnail of Post ID 1
* Reduced repetitive code with inheritance
* Compatibility check with WordPress 3.8

## 1.0

* Initial version


src/plugin.php
<?php
/*
Plugin Name: Custom Error Pages
Plugin URI: https://websistent.com/wordpress-plugins/custom-error-pages/
Description: Create custom 401 and 403 error pages with any WordPress theme without writing a single line of code, set it up and forget it.
Author: Jesin
Author URI: https://websistent.com
Version: 1.2
Text Domain: custom-error-pages
Domain Path: /languages
License: GPLv2
*/

if( !class_exists( 'Create_Custom_Error_Pages' ) && !class_exists( 'Custom_Error_Pages_Plugin' ) )
{
	class Custom_Error_Pages_Plugin
	{
		var $options;
		var $basename;
		var $slug;
		var $defaults;

		function __construct()
		{
			$this->basename = plugin_basename( __FILE__ );
			$this->slug = str_replace( array( basename( __FILE__ ), '/' ), '', $this->basename );
			$this->options = get_option( 'jesin_' . $this->slug );

			//Set the default text to display on error pages
			$this->defaults = array();
			$this->defaults['title_401'] = __( 'HTTP 401 Unauthorized', $this->slug );
			$this->defaults['content_401'] = __( 'Access denied due to invalid HTTP credentials', $this->slug );
			$this->defaults['title_403'] = __( 'HTTP 403 Forbidden', $this->slug );
			$this->defaults['content_403'] = __( 'You don\'t have permission to access this resource', $this->slug );

			add_action( 'init', array( $this, 'plugin_init' ) );
			add_filter( 'query_vars', array( $this, 'add_status_query_var' ) ); //Add a custom query variable
			add_action( 'parse_request', array( $this, 'get_status_query_var' ) ); //Check if the custom query variable exists
			add_filter( "comments_template", array( $this, 'plugin_disable_comments' ) ); //Disable comments for the generated "Virtual Page"
		}

		//Initialize the plugin
		function plugin_init()
		{
			load_plugin_textdomain( $this->slug, FALSE, $this->slug . '/languages' );		
		}

		//Adds custom "status" query variable
		function add_status_query_var( $vars )
		{
			$vars[] = 'status';
			return $vars;
		}

		//Checks for the existence of "status" query variable
		function get_status_query_var( &$wp )
		{
			if( isset( $wp->query_vars[ 'status' ] ) && ( $wp->query_vars[ 'status' ] == '401' || $wp->query_vars[ 'status' ] == '403' ) )
			{
				//Tells caching plugins like W3 Total Cache and WP Supercache not to cache these custom error pages
				define( 'DONOTCACHEPAGE', TRUE );

				$status = (int) $wp->query_vars[ 'status' ];

				//Cancel the canonical URL redirect
				remove_action( 'template_redirect', 'redirect_canonical' );

				//Create a custom error page based on our custom query variable
				$create_custom_error_pages = new Create_Custom_Error_Pages( array( 'options' => $this->options, 'code' => $status ) );
			}
		}
		
		//Disable comments for the generated page.
		//This is necessary since get_comments() with $post_id = 0 will return ALL comments of the current WordPress database.
		function plugin_disable_comments( $comment_template ) {
			global $post;
			if( $post->ID == 0 ) {
				return dirname( __FILE__ ) . '/comments_disabler.php';
			}
		}
	}

	//The Class which creates a "Virtual Page"
	class Create_Custom_Error_Pages
	{
		var $http_code;
		var $options;

		function __construct( $args )
		{
			$this->http_code = $args['code'];
			$this->options = $args['options'];

			add_action( 'wp', array( $this, 'send_status' ) );
			add_filter( 'the_posts', array( $this, 'generate_error_page' ) );
		}

		//Sends sets the appropriate HTTP error code in the response header
		function send_status()
		{
			status_header( $this->http_code );
		}

		//Create a Virtual "Page" with comments disabled on the fly
		function generate_error_page( $posts )
		{
			global $wp, $wp_query, $custom_error_pages_plugin;

			if( $wp_query->post_count == 0 )
			{
				$post = new stdClass;
				$post->ID = 0;
				$post->post_author = 1;
				$post->post_date = current_time( 'mysql' );
				$post->post_date_gmt = current_time( 'mysql', 1 );
				$post->post_content = ( (isset( $this->options['content_' . $this->http_code] ) && !empty( $this->options['content_' . $this->http_code] )) ? $this->options['content_' . $this->http_code] : $custom_error_pages_plugin->defaults['content_' . $this->http_code] );
				$post->post_title = ( (isset( $this->options['title_' . $this->http_code] ) && !empty( $this->options['title_' . $this->http_code] )) ? $this->options['title_' . $this->http_code] : $custom_error_pages_plugin->defaults['title_' . $this->http_code] );
				$post->post_excerpt = '';
				$post->post_status = 'publish';
				$post->comment_status = 'closed';
				$post->ping_status = 'closed';
				$post->post_password = '';
				$post->post_name = $this->http_code . '-error-' . time();
				$post->to_ping = '';
				$post->pinged = '';
				$post->post_type = 'page';
				$post->modified = $post->post_date;
				$post->modified_gmt = $post->post_date_gmt;
				$post->post_content_filtered = '';
				$post->post_parent = 0;
				$post->guid = get_home_url('/' . $post->post_name);
				$post->menu_order = 0;
				$post->post_tyle = 'page';
				$post->post_mime_type = '';
				$post->comment_count = 0;
 
				$posts = array( $post );
 
				$wp_query->is_page = TRUE;
				$wp_query->is_singular = TRUE;
				$wp_query->is_single = FALSE;
				$wp_query->is_home = FALSE;
				$wp_query->is_archive = FALSE;
				$wp_query->is_category = FALSE;
				unset( $wp_query->query['error'] );
				$wp_query->query_vars['error'] = '';
				$wp_query->is_404 = FALSE;
				$wp_query->comment_count = 0;
				$wp_query->current_comment = null;
			}
						
			return ( $posts );
		}
	}

	//Execution of the plugin begins here
	$custom_error_pages_plugin = new Custom_Error_Pages_Plugin;
}

if( is_admin() )
	require_once dirname( __FILE__ ) . '/admin_options.php';
